### PR TITLE
JSON-RPC e2e test POC

### LINKED ISSUE #19341: Update current PoC PR
See WIP PoC [here](https://github.com/MetaMask/metamask-extension/pull/18213)

**Acceptance criteria**
- Adding one example of a test that has a pop-up. That case should include:
  - Confirm case
  - Reject case
- Rebasing.

### PR DESCRIPTION
**Pull request by PeterYinusa**
## Explanation

Fixes: #19341 

This PR is a proof of concept for how we can test the JSON-RPC API.

Notes:

Currently, we are using ganache v7.0.4, the latest version is v7.9.0 at the time of writing.
If we wanted to test methods like `eth_feeHistory` for example, which is available from v7.5.0, we would need to upgrade Ganache.

Some methods like `eth_sendTransaction` initiate the MetaMask popup. In cases like this, the pattern of the test is slightly different as we need to interact with the popup

### After

https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/51700/workflows/255e5e8c-203b-485e-bbd9-840b2a713392/jobs/1505215/parallel-runs/0/steps/0-106

```
<testsuite name="Mocha Tests" tests="1" failures="0" errors="0" skipped="0" timestamp="Wed, 19 Jul 2023 11:17:10 GMT" time="16.69">
<testcase classname="eth_call" name="executes a new message call" time="16.688"/>
</testsuite>
<testsuite name="Mocha Tests" tests="1" failures="0" errors="0" skipped="0" timestamp="Wed, 19 Jul 2023 11:17:27 GMT" time="6.782">
<testcase classname="eth_chainId" name="returns the chain ID of the current network" time="6.78"/>
</testsuite>
<testsuite name="Mocha Tests" tests="2" failures="0" errors="0" skipped="0" timestamp="Wed, 19 Jul 2023 11:17:51 GMT" time="13.739">
<testcase classname="eth_sendTransaction" name="confirms a new transaction" time="6.839"/>
<testcase classname="eth_sendTransaction" name="rejects a new transaction" time="6.898"/>
</testsuite>

```

## Manual Testing Steps

1. yarn
2. yarn build:test
3. yarn test:e2e:chrome:rpc

## Pre-merge author checklist

- [ ] I've clearly explained:
  - [ ] What problem this PR is solving
  - [ ] How this problem was solved
  - [ ] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

**Comment by PeterYinusa** [comment]
@shanejonas @BelfordZ

File: package.json [hunk 1]
```diff
@@ -32,6 +32,7 @@
     "test:unit:mocha": "node ./test/run-unit-tests.js --mocha",
     "test:e2e:chrome": "SELENIUM_BROWSER=chrome node test/e2e/run-all.js",
     "test:e2e:chrome:snaps": "SELENIUM_BROWSER=chrome node test/e2e/run-all.js --snaps",
+    "test:e2e:chrome:rpc": "SELENIUM_BROWSER=chrome node test/e2e/run-all.js --rpc",
```
Comment (PeterYinusa) [inline package.json:35]: This command runs all the tests in the `json-rpc` folder

File: .circleci/config.yml [hunk 2]
```diff
@@ -699,6 +702,43 @@ jobs:
           path: test-artifacts
           destination: test-artifacts
 
+  test-e2e-chrome-rpc:
```
Comment (PeterYinusa) [inline .circleci/config.yml:705]: A separate CI job for the JSON-RPC tests. I chose Chrome as it's a faster browser to launch and run the tests than Firefox. I don't see any reason why we would need to run these tests cross browsers

File: test/e2e/json-rpc/eth_call.spec.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  };
+  it('executes a new message call', async function () {
+    await withFixtures(
+      {
+        dapp: true,
+        fixtures: new FixtureBuilder()
+          .withPermissionControllerConnectedToTestDapp()
+          .build(),
+        ganacheOptions,
+        smartContract,
```
Comment (PeterYinusa) [inline test/e2e/json-rpc/eth_call.spec.js:26]: This deploys a smart contract available in our test dapp

File: test/e2e/json-rpc/eth_call.spec.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                '000000000000000000000000' +
+                `${walletAddress.substring(2)}`,
+            },
+            'latest',
+          ],
+          id: 0,
+        });
+        const result = await driver.executeScript(
+          `return window.ethereum.request(${request})`,
+        );
```
Comment (PeterYinusa) [inline test/e2e/json-rpc/eth_call.spec.js:58]: Here we execute the `eth_call` method

File: test/e2e/json-rpc/eth_call.spec.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+        assert.equal(
+          result,
+          '0x0000000000000000000000000000000000000000000000000000000000000001',
+        );
```
Comment (PeterYinusa) [inline test/e2e/json-rpc/eth_call.spec.js:63]: This wallet owns one NFT

File: test/e2e/json-rpc/eth_call.spec.js (hunk 3 above)
Comment (PeterYinusa) [reply test/e2e/json-rpc/eth_call.spec.js:26]: In this example we deploy an ERC721 and mint a token

File: test/e2e/json-rpc/eth_call.spec.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+          jsonrpc: '2.0',
+          method: 'eth_call',
+          params: [
+            {
+              to: `${contract}`,
+              data:
+                `${balanceOf.slice(0, 10)}` +
+                '000000000000000000000000' +
+                `${walletAddress.substring(2)}`,
+            },
```
Comment (PeterYinusa) [inline test/e2e/json-rpc/eth_call.spec.js:51]: Call the smart contract method `balanceOf(address)`

File: test/e2e/json-rpc/eth_call.spec.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      },
+    ],
+  };
+  it('executes a new message call', async function () {
+    await withFixtures(
+      {
+        dapp: true,
+        fixtures: new FixtureBuilder()
+          .withPermissionControllerConnectedToTestDapp()
+          .build(),
```
Comment (PeterYinusa) [inline test/e2e/json-rpc/eth_call.spec.js:24]: This fixture allows us to connect to the test dapp with an existing connection

File: test/e2e/json-rpc/eth_call.spec.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        title: this.test.title,
+      },
+      async ({ driver, _, contractRegistry }) => {
+        const contract = contractRegistry.getContractAddress(smartContract);
+        await driver.navigate();
+        await driver.fill('#password', 'correct horse battery staple');
+        await driver.press('#password', driver.Key.ENTER);
+
+        // eth_call
+        await driver.openNewPage(`http://127.0.0.1:8080/?contract=${contract}`);
```
Comment (PeterYinusa) [inline test/e2e/json-rpc/eth_call.spec.js:36]: Navigate to the test dapp, so we can use `window.ethereum`

File: test/e2e/json-rpc/eth_call.spec.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        title: this.test.title,
+      },
+      async ({ driver, _, contractRegistry }) => {
+        const contract = contractRegistry.getContractAddress(smartContract);
+        await driver.navigate();
+        await driver.fill('#password', 'correct horse battery staple');
+        await driver.press('#password', driver.Key.ENTER);
+
+        // eth_call
+        await driver.openNewPage(`http://127.0.0.1:8080`);
```
Comment (PeterYinusa) [inline test/e2e/json-rpc/eth_call.spec.js:36]: Navigate to the test dapp, so we can use `window.ethereum`

File: test/e2e/json-rpc/eth_sendTransaction.spec.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+        // confirm transaction in mm popup
+        await driver.waitUntilXWindowHandles(3);
+        await driver.switchToWindowWithTitle('MetaMask Notification');
+        await driver.clickElement({ text: 'Confirm', tag: 'button' });
+        await driver.switchToWindowWithTitle('E2E Test Dapp');
+        const actualHash = await driver.executeScript(
+          `return window.transactionHash;`,
+        );
+        assert.equal(actualHash, expectedHash);
```
Comment (PeterYinusa) [inline test/e2e/json-rpc/eth_sendTransaction.spec.js:61]: When we confirm the transaction, the API returns a hash

File: test/e2e/json-rpc/eth_sendTransaction.spec.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+          .executeScript(`return window.transactionHash;`)
+          .then((data) => {
+            return data;
+          })
+          .catch((err) => {
+            return err;
+          });
+        assert.ok(
+          result.message.includes(
+            'MetaMask Tx Signature: User denied transaction signature.',
```
Comment (PeterYinusa) [inline test/e2e/json-rpc/eth_sendTransaction.spec.js:115]: When we reject the transaction, the API returns a MetaMask RPC Error: MetaMask Tx Signature: User denied transaction signature

**Review by pedronfigueiredo — APPROVED** [review]

**Comment by pedronfigueiredo** [comment]
LGTM

**Review by vthomas13 — APPROVED** [review]

**Review by brad-decker — APPROVED** [review]

**Pull request merged**